client/document.go

```diff
@@ -119,15 +117,274 @@ func NewDocFromMap(data map[string]any) (*Document, error) {
 	return doc, nil
 }
 
+var jsonArrayPattern = regexp.MustCompile(`^\s*\[.*\]\s*$`)
+
+// IsJSONArray returns true if the given byte array is a JSON Array.
+func IsJSONArray(obj []byte) bool {
+	return jsonArrayPattern.Match(obj)
+}
+
 // NewFromJSON creates a new instance of a Document from a raw JSON object byte array.
-func NewDocFromJSON(obj []byte) (*Document, error) {
-	data := make(map[string]any)
-	err := json.Unmarshal(obj, &data)
+func NewDocFromJSON(obj []byte, sd SchemaDescription) (*Document, error) {
+	doc := newEmptyDoc(sd)
+	err := doc.SetWithJSON(obj)
 	if err != nil {
 		return nil, err
 	}
+	err = doc.generateAndSetDocKey()
+	if err != nil {
+		return nil, err
+	}
+	return doc, nil
+}
+
+// ManyFromJSON creates a new slice of Documents from a raw JSON array byte array.
+func NewDocsFromJSON(obj []byte, sd SchemaDescription) ([]*Document, error) {
+	v, err := fastjson.ParseBytes(obj)
+	if err != nil {
+		return nil, err
+	}
+	a, err := v.Array()
+	if err != nil {
+		return nil, err
+	}
+
+	docs := make([]*Document, len(a))
+	for _, v := range a {
+		o, err := v.Object()
+		if err != nil {
+			return nil, err
+		}
+		doc := newEmptyDoc(sd)
+		err = doc.setWithFastJSONObject(o)
+		if err != nil {
+			return nil, err
+		}
+		err = doc.generateAndSetDocKey()
+		if err != nil {
+			return nil, err
+		}
+		docs = append(docs, doc)
+	}
+
+	return docs, nil
+}
+
+// validateFieldSchema takes a given value as an interface,
+// and ensures it matches the supplied field description.
+// It will do any minor parsing, like dates, and return
+// the typed value again as an interface.
+func validateFieldSchema(val any, field FieldDescription) (any, error) {
```

Comment:
note: I've very quickly skimmed of this and it's child-funcs as I assume it is copy-pasted from elsewhere in our codebase without any real changes - do give me a shout if you want it reviewed.